src/gluonts/dataset/arrow.py

```diff
@@ -0,0 +1,312 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+import re
+from pathlib import Path
+from typing import (
+    Union,
+    Dict,
+    List,
+    Any,
+    cast,
+    Optional,
+    Tuple,
+    Iterator,
+    Iterable,
+)
+
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from gluonts.dataset import util
+from gluonts.dataset.util import get_bounds_for_mp_data_loading
+from gluonts.gluonts_tqdm import tqdm
+
+from .common import Dataset, DataEntry, ProcessDataEntry, SourceContext
+
+_ARROW_NUMPY_SHAPE_SUFFIX = "._np_shape"
+
+
+class ArrowWriter:
+    """
+    Write records to an arrow table. The first record is used to infer the
+    schema. Lists are converted to numpy int or float arrays if possible.
+    Strings of the form 'yyyy-mm-dd ...' are converted to timestamps if
+    possible.
+
+    Note:
+    Numpy arrays with dimension > 1 are stored in two columns one with a
+    1d flat data an additional column called "<array_name>._np_shape" that
+    stores the array shape. One could also do this via pyarrow/pandas
+    extension types, but it seemed overly complicated for this use case.
+    """
+
+    def __init__(
```

Comment:
Couldn't we use a dataclass for this? `pydantic` offers a datclass which would also convert the `Path` automatically.